feat(swarm)!: allow `NetworkBehaviour`s to manage connections

thomaseizinger opened the pull request:
## Description

<!-- Please write a summary of your changes and why you made them.-->
<!-- This section will appear as the commit message after merging. Please craft it accordingly. -->

Previously, a `ConnectionHandler` was immediately requested from the `NetworkBehaviour` as soon as a new dial was initiated or a new incoming connection accepted.

With this patch, we delay the creation of the handler until the connection is actually established and fully upgraded, i.e authenticated and multiplexed.

As a consequence, `NetworkBehaviour::new_handler` is now deprecated in favor of a new set of callbacks:

- `NetworkBehaviour::handle_pending_inbound_connection`
- `NetworkBehaviour::handle_pending_outbound_connection`
- `NetworkBehaviour::handle_established_inbound_connection`
- `NetworkBehaviour::handle_established_outbound_connection`

All callbacks are fallible, allowing the `NetworkBehaviour` to abort the connection either immediately or after it is fully established. All callbacks also receive a `ConnectionId` parameter which uniquely identifies the connection. For example, in case a `NetworkBehaviour` issues a dial via `NetworkBehaviourAction::Dial`, it can unambiguously detect this dial in these lifecycle callbacks via the `ConnectionId`. 

Finally, `NetworkBehaviour::handle_pending_outbound_connection` also replaces `NetworkBehaviour::addresses_of_peer` by allowing the behaviour to return more addresses to be used for the dial.

Resolves #2824.

## Notes

<!-- Any notes or remarks you'd like to make about the PR. -->

This is a variant of https://github.com/libp2p/rust-libp2p/pull/3099 with less breaking changes and more deprecations.

- Depends-On: https://github.com/libp2p/rust-libp2p/pull/3243
- Depends-On: https://github.com/libp2p/rust-libp2p/pull/3239
- Depends-On: https://github.com/libp2p/rust-libp2p/pull/3253
- Depends-On: https://github.com/libp2p/rust-libp2p/pull/3272
- Depends-On: https://github.com/libp2p/rust-libp2p/pull/3264
- Depends-On: https://github.com/libp2p/rust-libp2p/pull/3346
- Depends-On: https://github.com/libp2p/rust-libp2p/pull/3334
- Depends-On: https://github.com/libp2p/rust-libp2p/pull/3327
- Depends-On: https://github.com/libp2p/rust-libp2p/pull/3335
- Depends-On: https://github.com/libp2p/rust-libp2p/pull/3372
- Depends-On: https://github.com/libp2p/rust-libp2p/pull/3328
- Depends-On: https://github.com/libp2p/rust-libp2p/pull/3368
- Depends-On: https://github.com/libp2p/rust-libp2p/pull/3373
- Depends-On: https://github.com/libp2p/rust-libp2p/pull/3374
- Depends-On: https://github.com/libp2p/rust-libp2p/pull/3375
- Depends-On: https://github.com/libp2p/rust-libp2p/pull/3381
- Depends-On: https://github.com/libp2p/rust-libp2p/pull/3384


## Links to any relevant issues

<!-- Reference any related issues.-->

## Open Questions

<!-- Unresolved questions, if any. -->

- ~The `ConnectionId` is currently generated through a random number generator. This is not ideal and could in theory lead to collisions:~
	- ~Should we use a UUID instead? I think those are less likely to produce a collision.~
	- ~Should we manage the connection IDs centrally, i.e. in the `Swarm`? This would require us to provide a facility to the `NetworkBehaviour` in how it can generate a new one to ensure it can track state for its own dial. We could use `PollParameters` for that. One downside of this is a slight ergonomic hit because the user can only fully construct a `NetworkBehaviourAction::Dial` within `NetworkBehaviour::poll` and not within the other callbacks. Some of our protocols already do that via an `ActionBuilder` pattern.~

## Change checklist

<!-- Please add a Changelog entry in the appropriate crates and bump the crate versions if needed. See <https://github.com/libp2p/rust-libp2p/blob/master/docs/release.md#development-between-releases>-->

- [ ] I have performed a self-review of my own code
- [x] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] A changelog entry has been made in the appropriate crates

thomaseizinger committed:
Use GATs to avoid lifetimes on trait interface

thomaseizinger committed:
Bump MSRV accordingly

thomaseizinger committed:
Bump MSRV of libp2p-metrics

thomaseizinger committed:
Merge branch 'master' into feat-kademlia/use-gats

thomaseizinger committed:
Refactor dialling to have a more linear flow

thomaseizinger committed:
Introduce new callbacks for connection management

thomaseizinger committed:
Fix up some more code

thomaseizinger committed:
Complete deprecation notice

thomaseizinger committed:
Fix some more errors

thomaseizinger committed:
Refine deprecation message

thomaseizinger committed:
Delegate to `addresses_of_peer` by default

thomaseizinger committed:
Add hack to clear listen addresses

thomaseizinger committed:
Merge branch 'master' into feat-kademlia/use-gats

thomaseizinger committed:
Merge branch 'feat-kademlia/use-gats' of github.com:libp2p/rust-libp2p into feat-kademlia/use-gats

thomaseizinger committed:
Add changelog entry

thomaseizinger committed:
Merge branch 'master' into 2824-deprecate-into-connection-handler

thomaseizinger committed:
Use an atomic counter to avoid collisions on connection IDs

thomaseizinger committed:
Fix minor compile error

thomaseizinger committed:
Fill in a few more bits

thomaseizinger committed:
Don't be generic over transport

thomaseizinger committed:
Remove type parameter for transport error

thomaseizinger committed:
Remove unnecessary type hint

thomaseizinger committed:
Remove type parameters from `ConcurrentDial`

thomaseizinger committed:
Fix rustfmt

thomaseizinger committed:
Merge branch 'feat-kademlia/use-gats' into 2824-deprecate-into-connection-handler

thomaseizinger committed:
Merge branch '2824-simplify-dial-opts' into 2824-deprecate-into-connection-handler

thomaseizinger committed:
Merge branch '2824-no-generic-transport-pool' into 2824-deprecate-into-connection-handler

thomaseizinger committed:
Remove `PendingConnectionDenied`

This is handled by `DialError` and `ListenFailure`.

thomaseizinger committed:
Handle new DialError variants

thomaseizinger committed:
Make `libp2p-kad` compile

thomaseizinger committed:
Fix a few more errors

thomaseizinger committed:
No more compile errors!

mxinden commented on swarm/src/behaviour.rs:
````diff
@@ -123,37 +116,133 @@ pub(crate) type THandlerOutEvent<THandler> =
 /// }
 /// ```
-pub trait NetworkBehaviour: 'static {
+pub trait NetworkBehaviour: Sized + 'static {
````
How come this has to be explicitly `Sized` now?

thomaseizinger replied:
I don't remember, let me try remove it and see what breaks!

thomaseizinger replied:
Not needed!

mxinden commented on swarm/src/behaviour.rs:
```diff
@@ -123,37 +116,133 @@ pub(crate) type THandlerOutEvent<THandler> =
+    /// We have access to:
+    ///
+    /// - The [`PeerId`], if known. Remember that we can dial without a [`PeerId`].
+    /// - All addresses passes to [`DialOpts`] are passed in here too.
```
```suggestion
    /// - All addresses passed to [`DialOpts`] are passed in here too.
```

mxinden commented on swarm/src/lib.rs:
```diff
@@ -500,146 +511,100 @@ where
+                addresses_from_opts.as_slice(),
+                dial_opts.role_override(),
+                connection_id,
+            ) {
```
Just calling this out to make sure I am not missing something. I don't think this is problematic.

Addresses collected through `handle_pending_outbound_connection` might not actually be used in the dial in case the dialer did not enable `extend_addresses_through_behaviour`.

thomaseizinger replied:
Yep, I think this might warrant a `debug` log.

elenaf9 replied:
Could we have `handle_pending_outbound_connection` return something like `impl Iterator<Item = &Multiaddr>` instead? That way we'd avoid unnecessary cloning of addresses if they end up not being used anyway.

thomaseizinger replied:
> Could we have `handle_pending_outbound_connection` return something like `impl Iterator<Item = &Multiaddr>` instead? That way we'd avoid unnecessary cloning of addresses if they end up not being used anyway.

I'll explore it, good idea!

mxinden replied:
> That way we'd avoid unnecessary cloning of addresses if they end up not being used anyway.

My initial comment was not aimed at the potential performance impact, but rather at the surprise for an implementer of the API, expecting their addresses to be used in the final dial. Instead of optimizing away the allocations, I suggest adding a comment to the trait's method to prevent implementers being surprised.

thomaseizinger replied:
Are we happy with the log statement that I added?

mxinden replied:
Happy!

mxinden commented on swarm/src/behaviour.rs:
```diff
@@ -123,37 +116,133 @@ pub(crate) type THandlerOutEvent<THandler> =
+    /// At this point, we have verified their [`PeerId`] and we know, which particular [`Multiaddr`] succeeded in the dial.
+    /// In order to actually use this connection, this function must return a [`ConnectionHandler`].
+    /// Returning an error will immediately close the connection.
+    fn handle_established_outbound_connection(
```
Checking my understanding, we still need the `FromSwarm::ConnectionEstablished` event as any `NetworkBehaviour` within the hierarchy of behaviours might deny the connection.

If so, what do you think of elaborating this in the methods doc comment? Would that confuse the user more than it helps?

thomaseizinger replied:
Yes, the event indicates that the connection was not denied and can actually be used.

mxinden commented on swarm/src/behaviour.rs:
```diff
@@ -123,37 +116,133 @@ pub(crate) type THandlerOutEvent<THandler> =
     /// The addresses will be tried in the order returned by this function, which means that they
     /// should be ordered by decreasing likelihood of reachability. In other words, the first
     /// address should be the most likely to be reachable.
+    #[deprecated(note = "Use `NetworkBehaviour::handle_pending_outbound_connection` instead.")]
```
I like that this is combined with the connection management method now.

mxinden commented on swarm/src/behaviour.rs:
```diff
@@ -270,14 +358,10 @@ pub trait NetworkBehaviour: 'static {
         peer_id: Option<PeerId>,
-        handler: Self::ConnectionHandler,
         error: &DialError,
+        id: ConnectionId,
```
I am assuming that this might simplify a couple of `NetworkBehaviour` implementations. :+1:

thomaseizinger replied:
To be honest, these events are quite a mess, esp. the divergence of `FromSwarm` and `SwarmEvent`. It would be nice if we can get rid of `SwarmEvent` apart from it emitting behaviour events.

mxinden reviewed (commented):
Well thought through. Thanks @thomaseizinger!

I reviewed `swarm/src/lib.rs` and `swarm/src/behaviour.rs` under the assumption that the other files are not yet ready for review. Please ping me in case I should look at more or take another look at the above.

thomaseizinger committed:
Merge branch 'master' into 2824-deprecate-into-connection-handler

thomaseizinger committed:
Undo explicit `Sized` bound

thomaseizinger committed:
Let user know that we are discarding addresses

thomaseizinger committed:
Fix unused variable warning

thomaseizinger committed:
Fix unused import warnings

thomaseizinger committed:
Port `dcutr` protocol

thomaseizinger committed:
Re-implement relay

thomaseizinger committed:
Fix some deprecation warnings

thomaseizinger committed:
Fix some deprecation warnings kad

thomaseizinger committed:
Port autonat

thomaseizinger committed:
Port identify

thomaseizinger committed:
Fix more deprecation warnings

thomaseizinger committed:
Ignore clippy warning

thomaseizinger committed:
Fix derive and remove todo

thomaseizinger committed:
Make derive tests compile by using type alias

thomaseizinger committed:
Ignore clippy warnings in generated code

thomaseizinger committed:
Fix more warnings by allowing deprecations

thomaseizinger committed:
Merge branch 'master' into 2824-deprecate-into-connection-handler

thomaseizinger commented on protocols/relay/src/v2/client.rs:
```diff
@@ -146,11 +152,51 @@ impl Client {
+        local_addr: &Multiaddr,
+        remote_addr: &Multiaddr,
+    ) -> Result<THandler<Self>, Box<dyn std::error::Error + Send + 'static>> {
+        let is_relayed = local_addr.iter().any(|p| p == Protocol::P2pCircuit); // TODO: Make this an extension on `Multiaddr`.
```
TODO: This should be in a separate PR.

thomaseizinger commented on protocols/ping/src/lib.rs:
```diff
@@ -166,4 +160,33 @@ impl NetworkBehaviour for Behaviour {
+        &mut self,
+        peer: PeerId,
+        _: ConnectionId,
+        result: THandlerOutEvent<Self>,
```
TODO: Using these type aliases everywhere should be a separate PR.

thomaseizinger commented on protocols/relay/src/v2/client.rs:
```diff
@@ -304,20 +349,21 @@ impl NetworkBehaviour for Client {
+                        let (action, connection_id) = NetworkBehaviourAction::dial(
+                            DialOpts::peer_id(relay_peer_id)
+                                .addresses(vec![relay_addr])
+                                .extend_addresses_through_behaviour() // TODO: Why?
```
TODO: Fixing this should be a separate PR.

thomaseizinger commented on protocols/relay/src/v2/relay/handler.rs:
```diff
@@ -432,6 +394,22 @@ pub struct Handler {
+            alive_lend_out_substreams: Default::default(),
+            circuits: Default::default(),
+        }
+    }
```
TODO: Introducing this can be done in a separate PR.

thomaseizinger commented on protocols/rendezvous/src/client.rs:
```diff
@@ -173,26 +169,74 @@ impl NetworkBehaviour for Behaviour {
             .flatten()
             .cloned()
-            .collect()
+            .collect())
```
TODO: Attempt to minimize diff here be storing addresses in separate variable.

thomaseizinger commented on protocols/rendezvous/src/client.rs:
```diff
@@ -173,26 +169,74 @@ impl NetworkBehaviour for Behaviour {
+            | FromSwarm::ListenerClosed(_)
+            | FromSwarm::NewExternalAddr(_)
+            | FromSwarm::ExpiredExternalAddr(_) => {}
+        }
```
TODO: Revert reordering this to minimize diff.

thomaseizinger commented on swarm/src/connection/pool.rs:
```diff
@@ -285,16 +293,14 @@ pub enum PoolEvent<THandler: IntoConnectionHandler> {
         peer_id: PeerId,
         /// The produced event.
-        event: THandlerOutEvent<THandler>,
+        event: THandler::OutEvent,
```
TODO: Revert this.

thomaseizinger commented on swarm/src/connection/pool.rs:
```diff
@@ -310,11 +316,12 @@ pub enum PoolEvent<THandler: IntoConnectionHandler> {
     pub fn new(local_id: PeerId, config: PoolConfig, limits: ConnectionLimits) -> Self {
-        let (pending_connection_events_tx, pending_connection_events_rx) = mpsc::channel(0);
+        let (pending_connection_events_tx, pending_connection_events_rx) =
+            mpsc::channel(config.task_event_buffer_size);
```
TODO: Bad merge.

thomaseizinger commented on swarm/src/connection/pool.rs:
```diff
@@ -348,7 +354,7 @@ where
         &mut self,
         id: ConnectionId,
-    ) -> Option<&mut EstablishedConnection<THandlerInEvent<THandler>>> {
+    ) -> Option<&mut EstablishedConnection<THandler::InEvent>> {
```
TODO: Revert this.

thomaseizinger commented on swarm/src/connection/pool.rs:
```diff
@@ -774,23 +787,23 @@ where
                         concurrent_dial_errors,
-                        established_in,
+                        established_in: accepted_at.elapsed(),
+                        supported_protocols,
```
TODO: Minimize diff here.

thomaseizinger commented on swarm/src/connection/pool.rs:
```diff
@@ -774,23 +787,23 @@ where
                         endpoint,
                         abort_notifier: _,
-                        accepted_at: _, // Ignoring the time it took for the connection to fail.
+                        accepted_at: _,
```
TODO: Minimize diff here.

thomaseizinger commented on swarm/src/connection/pool.rs:
```diff
@@ -1095,6 +1106,12 @@ impl PoolConfig {
+    pub fn with_executor(mut self, executor: Box<dyn Executor + Send>) -> Self {
+        self.executor = Some(executor);
+        self
+    }
```
TODO: Bad merge?

thomaseizinger commented on swarm/src/connection/pool/task.rs:
```diff
@@ -66,7 +66,10 @@ pub enum PendingConnectionEvent {
+pub enum EstablishedConnectionEvent<THandler>
+where
+    THandler: ConnectionHandler,
+{
```
TODO: Revert or send separate PR.

thomaseizinger commented on swarm/src/dummy.rs:
```diff
@@ -17,20 +22,24 @@ impl NetworkBehaviour for Behaviour {
+        _: ConnectionId,
+    ) -> Result<THandler<Self>, Box<dyn Error + Send + 'static>> {
+        Ok(ConnectionHandler)
     }
```
TODO: Revert reordering or send separate PR.

thomaseizinger commented on core/src/connection.rs:
```diff
@@ -19,27 +19,37 @@
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "{}", self.0)
     }
 }
```
TODO: Wait for https://github.com/libp2p/rust-libp2p/pull/3221 to be merged here.

thomaseizinger committed:
Remove unnecessary `extern crate`

thomaseizinger commented on protocols/floodsub/src/layer.rs:
```diff
@@ -344,16 +336,31 @@ impl NetworkBehaviour for Floodsub {
+        _: &Multiaddr,
+        _: Endpoint,
+        _: ConnectionId,
+    ) -> std::result::Result<THandler<Self>, Box<dyn std::error::Error + Send + 'static>> {
```
TODO: Audit PR for unnecessarily qualified `std::result::Result`.

thomaseizinger commented on protocols/relay/src/v2/client.rs:
```diff
@@ -188,9 +234,9 @@ impl NetworkBehaviour for Client {
+        _event: THandlerOutEvent<Self>,
     ) {
-        let handler_event = match handler_event {
+        let handler_event = match _event {
```
TODO: Minimize diff.

thomaseizinger commented on protocols/relay/src/v2/client/handler.rs:
```diff
@@ -213,6 +155,21 @@ pub struct Handler {
+            circuit_deny_futs: Default::default(),
+            send_error_futs: Default::default(),
+        }
+    }
```
TOOD: Can be in a separate PR to reduce diff.

thomaseizinger commented on protocols/rendezvous/src/client.rs:
```diff
@@ -31,23 +31,19 @@ use instant::Duration;
-            Event,
-            SubstreamConnectionHandler<void::Void, outbound::Stream, outbound::OpenInfo>,
-        >,
-    >,
```
TODO: Reduce diff here by first migrating everything to `THandler`.

thomaseizinger committed:
Revert let-else usage

thomaseizinger committed:
Less let else

thomaseizinger committed:
No let else in mdns

thomaseizinger committed:
No let-else in dcutr

thomaseizinger committed:
No let else in rendezvous

thomaseizinger committed:
Last let-else usage

thomaseizinger reviewed (commented):
A few remarks for myself.

thomaseizinger commented on swarm/src/lib.rs:
```diff
@@ -917,25 +966,42 @@ where
+                    Err(error) => {
+                        // TODO: self.behaviour.on_swarm_event(FromSwarm::ListenFailure())
+
+                        return Some(todo!("Report connection error event, these are currently modeled in a weird way where we split them by incoming/outgoing on the `SwarmEvent` level but everywhere else by established/pending"));
```
@mxinden: Big question mark here on how we want to model this. Input welcome.

mxinden replied:
If I understand correctly, we didn't return a `SwarmEvent` in this particular case at all, right? I am not opposed to returning a `SwarmEvent` in the case were we reject a pending inbound connection.

I don't have a strong opinion on which of the two namespace splits should take preference. Striving for consistency of the two sounds good to me, though seems to be a separate undertaking.

thomaseizinger replied:
> If I understand correctly, we didn't return a `SwarmEvent` in this particular case at all, right?

Right, we didn't. I hadn't considered _not_ returning one, that might indeed unblock this bit, thanks!

> Striving for consistency of the two sounds good to me, though seems to be a separate undertaking.

It is a bit of a mess 😔

thomaseizinger replied:
> > If I understand correctly, we didn't return a `SwarmEvent` in this particular case at all, right?
> 
> Right, we didn't. I hadn't considered _not_ returning one, that might indeed unblock this bit, thanks!
> 
> > Striving for consistency of the two sounds good to me, though seems to be a separate undertaking.
> 
> It is a bit of a mess pensive

Actually, this still proves to be a bit difficult. In follow-up PRs to this, I think we should re-implement some of the current functionality in regards to banning, connection limits etc on top of this new infrastructure. That will remove several, currently explicit error cases. I think if we don't model these error properly, it will be difficult for users to understand what is happening.

I am proposing to align these errors before we merge this PR.

thomaseizinger replied:
> I am proposing to align these errors before we merge this PR.

It might have to be part of this, I'll see what I can do.

mxinden replied:
> I think we should re-implement some of the current functionality in regards to banning, connection limits etc on top of this new infrastructure.

Just to make sure we agree here, long term I would like all connection management to happen through `NetworkBehaviour` implementations. I.e. `libp2p-swarm` should not be aware of the concepts of connection limits nor peer banning.

mxinden replied:
> I am proposing to align these errors before we merge this PR.

Fine with me.

thomaseizinger replied:
> > I think we should re-implement some of the current functionality in regards to banning, connection limits etc on top of this new infrastructure.
> 
> Just to make sure we agree here, long term I would like all connection management to happen through `NetworkBehaviour` implementations. I.e. `libp2p-swarm` should not be aware of the concepts of connection limits nor peer banning.

Yes. I assumed we would still provide a simple `Blocklist` behaviour that replaces the current banning functionality?

mxinden replied:
:+1: we are on the same page. I am fine with `Blocklist` to live within `libp2p-swarm`, as long as it simply implements `NetworkBehaviour`.

mxinden commented on core/src/connection.rs:
```diff
@@ -19,27 +19,37 @@
 
-impl std::ops::Add<usize> for ConnectionId {
-    type Output = Self;
+impl Default for ConnectionId {
```
Ignoring the fact that there has already been a `ConnectionId::new` and that that method takes a parameter.

My intuition for `Default` would be that two `Default::default()` would equal. Is that idiomatic Rust or only my intuition.

This is not particularly important. Fine with keeping it as is.

thomaseizinger replied:
Good question. I think I agree with you, now that you are pointing it out!

We can easily replace it with a public `next_unused` constructor.

thomaseizinger committed:
Remove fn from bad merge

thomaseizinger committed:
Minimize diff

thomaseizinger committed:
Attempt to minimize diff

thomaseizinger committed:
Undo re-ordering to reduce diff

thomaseizinger committed:
Minimize diff (and fix bad merge)

thomaseizinger committed:
Remove default impl for `ConnectionId` in favor of `next`

thomaseizinger committed:
Deprecate `IntoMultiHandler`

thomaseizinger committed:
Implement dial failure reporting

thomaseizinger committed:
Fix formatting

mxinden commented on swarm/src/behaviour.rs:
````diff
@@ -437,200 +524,12 @@ pub trait PollParameters {
-    /// # struct PreciousMessage(String);
-    /// ```
-    Dial { opts: DialOpts, handler: THandler },
+    Dial { opts: DialOpts, id: ConnectionId },
````
A user could create a `NetworkBehaviourAction::Dial` with an already-in-use `ConnectionId`. Should we prevent this at compile time? Should one only be able to create a `NetworkBehaviourAction::Dial` with a **new** `ConnectionId` provided?

thomaseizinger replied:
It is a valid concern. Do you have a design in mind already? We would have to either make the entire enum and opaque type or at least create a `Dial` type with private fields so we can encapsulate it properly.

mxinden replied:
> Do you have a design in mind already?

No design in mind. I don't see this as a blocker.

thomaseizinger replied:
Note: In https://github.com/libp2p/rust-libp2p/pull/3328, we now use a design that together with https://github.com/libp2p/rust-libp2p/pull/3346 will allow us to make the constructor private and thus prevent at compile-time that users can assign an existing `ConnectionId` to a new connection.

mxinden reviewed (commented):
Needs changelog entries and doc comments. Otherwise this looks good to me.

Great work on making the upgrade path from `new_handler` and `addresses_of_peer` to the new methods easy! Double checked the flow through the various combinator `NetworkBehaviour` implementations.

thomaseizinger committed:
Use single quotes to avoid running commands

thomaseizinger committed:
Use intermediary variable

thomaseizinger committed:
Make connection IDs globally unique

thomaseizinger committed:
Deprecate old new function

thomaseizinger committed:
Add changelog entry

thomaseizinger committed:
Merge branch 'master' into no-run-title-as-command

thomaseizinger committed:
Remove `handler` field from `NetworkBehaviourAction::Dial`

thomaseizinger committed:
Port dcutr protocol

thomaseizinger committed:
Port relay protocol

thomaseizinger committed:
Add changelog entry to swarm

thomaseizinger committed:
Merge branch 'no-run-title-as-command' into 2824-unique-connection-ids

thomaseizinger committed:
Merge branch '2824-unique-connection-ids' into 2824-no-handler-in-dial

thomaseizinger committed:
Remove ordering functionality from `ConnectionId`

thomaseizinger committed:
Merge branch '2824-unique-connection-ids' of github.com:libp2p/rust-libp2p into 2824-unique-connection-ids

thomaseizinger committed:
Remove unnecessary visibility qualifiers

thomaseizinger committed:
Convert `DialOpts` to struct variant

thomaseizinger committed:
Eagerly set defaults in `DialOpts`

Instead of carrying around an enum and parsing its contents later,
directly set the correct values as we construct the final `DialOpts`.

thomaseizinger committed:
Merge branch '2824-refactor-dial-opts' into 2824-no-handler-in-dial

thomaseizinger committed:
Make `ConnectionId` part of `DialOpts`

thomaseizinger committed:
Inline `_dial`

thomaseizinger committed:
Rustfmt

thomaseizinger committed:
Fix wrong default

thomaseizinger committed:
Merge branch '2824-refactor-dial-opts' into 2824-no-handler-in-dial

thomaseizinger committed:
Merge branch 'master' into 2824-deprecate-into-connection-handler

thomaseizinger committed:
Merge branch 'master' into 2824-unique-connection-ids

thomaseizinger committed:
Merge branch '2824-unique-connection-ids' of github.com:libp2p/rust-libp2p into 2824-unique-connection-ids

thomaseizinger committed:
Merge branch 'master' into 2824-unique-connection-ids

thomaseizinger committed:
Merge branch '2824-unique-connection-ids' into 2824-no-handler-in-dial

thomaseizinger committed:
Expand changelog entry

thomaseizinger committed:
Move most important changelog entry to the top

thomaseizinger committed:
Add more text to changelog entry

thomaseizinger committed:
Fix formatting

thomaseizinger committed:
Merge branch 'master' into 2824-unique-connection-ids

thomaseizinger committed:
Merge branch '2824-unique-connection-ids' into 2824-no-handler-in-dial

thomaseizinger committed:
Merge branch '2824-no-handler-in-dial' into 2824-deprecate-into-connection-handler

thomaseizinger committed:
Fix most errors

mergify[bot] committed:
Merge branch 'master' into 2824-unique-connection-ids

thomaseizinger committed:
Merge branch 'master' into 2824-unique-connection-ids

thomaseizinger committed:
Merge branch '2824-unique-connection-ids' into 2824-no-handler-in-dial

thomaseizinger committed:
Merge branch '2824-no-handler-in-dial' into 2824-deprecate-into-connection-handler

thomaseizinger committed:
Fix implementation of dcutr

thomaseizinger committed:
Reduce diff

thomaseizinger committed:
Undo accidential change

thomaseizinger committed:
Reduce diff

thomaseizinger committed:
Attempt to reduce diff

thomaseizinger committed:
Merge branch 'master' into 2824-no-handler-in-dial

thomaseizinger committed:
Remove unnecessary `..`

thomaseizinger committed:
Merge branch '2824-no-handler-in-dial' into 2824-deprecate-into-connection-handler

thomaseizinger committed:
WIP

thomaseizinger committed:
Remove unused error variant

thomaseizinger committed:
Apply suggestions from code review

thomaseizinger committed:
Merge branch '2824-remove-unused-error' into 2824-introduce-listen-error

thomaseizinger committed:
WIP

thomaseizinger committed:
Remove unused `DialError::ConnectionIo` variant

thomaseizinger committed:
Merge branch '2824-remove-unused-error' into 2824-remove-unused-connection-error

thomaseizinger committed:
add changelog entry

thomaseizinger committed:
Fix missing colon

thomaseizinger committed:
Merge branch '2824-remove-unused-connection-error' into 2824-introduce-listen-error

thomaseizinger committed:
Remove unused `ConnectionIo` variant from `ListenError`

thomaseizinger committed:
Update docs

thomaseizinger committed:
Add changelog entry

thomaseizinger committed:
Update changelog

thomaseizinger committed:
Merge branch 'master' into 2824-no-handler-in-dial

thomaseizinger committed:
Merge branch '2824-introduce-listen-error' into 2824-deprecate-into-connection-handler

thomaseizinger committed:
Report cause of denied connection

thomaseizinger committed:
Introduce dedicated error type

thomaseizinger commented on protocols/ping/src/lib.rs:
```diff
@@ -167,4 +159,33 @@ impl NetworkBehaviour for Behaviour {
+        result: THandlerOutEvent<Self>,
+    ) {
+        self.events.push_front(Event { peer, result })
+    }
```
Note to self: Check that this diff gets better once we merge https://github.com/libp2p/rust-libp2p/pull/3368.

thomaseizinger committed:
Don't comment docs

thomaseizinger committed:
Fix docs

thomaseizinger commented on swarm/src/dummy.rs:
```diff
@@ -51,6 +57,23 @@ impl NetworkBehaviour for Behaviour {
+        _: &mut impl PollParameters,
+    ) -> Poll<NetworkBehaviourAction<Self::OutEvent, THandlerInEvent<Self>>> {
+        Poll::Pending
+    }
```
TODO: Check that this diff improves with https://github.com/libp2p/rust-libp2p/pull/3368.

thomaseizinger committed:
Migrate `CallTraceBehaviour`

thomaseizinger committed:
Temporarily allow usage of `new_handler`

thomaseizinger committed:
Refactor gossipsub config handler to avoid calls to trait functions

thomaseizinger committed:
Undo bad doc comment change

thomaseizinger committed:
Fix rustdocs

thomaseizinger committed:
Fix deprecation warnings

thomaseizinger committed:
Fix failing gossipsub test

thomaseizinger committed:
Fix more docs

thomaseizinger committed:
Fix beta clippy

thomaseizinger committed:
Actually fix beta clippy

thomaseizinger commented on protocols/dcutr/src/behaviour_impl.rs:
```diff
@@ -235,18 +232,93 @@ impl Behaviour {
+        local_addr: &Multiaddr,
+        remote_addr: &Multiaddr,
+    ) -> Result<THandler<Self>, ConnectionDenied> {
+        let is_relayed = local_addr.iter().any(|p| p == Protocol::P2pCircuit); // TODO: Make this an extension on `Multiaddr`.
```
Any thoughts on how to handle this? With the introduction of the new APIs, we are no longer passing a `ConnectedPoint` which has the `is_relayed` utility function on it.

mxinden replied:
I am fine with:

- Adding a method `Multiaddr::is_relayed`
- Adding a method `MultiaddrExt::is_relayed`
- Keeping as is, i.e. `iter().any`

thomaseizinger committed:
Resolve remaining todos

thomaseizinger committed:
Don't use else block if we return in the previous one

thomaseizinger committed:
Remove closure in favor of spawning connection from the outside

thomaseizinger committed:
Merge branch 'master' into 2824-no-handler-in-dial

thomaseizinger committed:
Add `ConnectionId` to `ListenFailure`

thomaseizinger committed:
Clean up pending commands on failure

thomaseizinger committed:
Add docs and fix bug

thomaseizinger committed:
Merge branch 'master' into 2824-no-handler-in-dial

thomaseizinger committed:
Consistently use `THandlerInEvent`

thomaseizinger committed:
Use type alias

thomaseizinger committed:
Remove unnecessary indentation and "quote"-ing

thomaseizinger committed:
Fix bad changelog entry

thomaseizinger committed:
Merge branch '2824-no-handler-in-dial' into 2824-deprecate-into-connection-handler

thomaseizinger committed:
Spawn connection at right point in time

thomaseizinger committed:
Attempt to reduce diff

thomaseizinger committed:
Reduce diff further

thomaseizinger committed:
Less diff ...

thomaseizinger committed:
Don't override deprecated callbacks anywhere

thomaseizinger committed:
Better order of callbacks

thomaseizinger committed:
Spawn connection at right point in time

Otherwise the num-established counter is wrong.

thomaseizinger committed:
Be consistent in order of parameters

thomaseizinger committed:
Merge branch '2824-no-handler-in-dial' into 2824-deprecate-into-connection-handler

thomaseizinger commented on protocols/relay/src/behaviour.rs:
```diff
@@ -250,17 +252,61 @@ impl Behaviour {
+        local_addr: &Multiaddr,
+        remote_addr: &Multiaddr,
+    ) -> Result<THandler<Self>, ConnectionDenied> {
+        let is_relayed = local_addr.iter().any(|p| p == Protocol::P2pCircuit); // TODO: Make this an extension on `Multiaddr`.
```
TODO: Resolve this in some form.

thomaseizinger commented on protocols/relay/src/priv_client.rs:
```diff
@@ -156,11 +159,51 @@ impl Behaviour {
+        local_addr: &Multiaddr,
+        remote_addr: &Multiaddr,
+    ) -> Result<THandler<Self>, ConnectionDenied> {
+        let is_relayed = local_addr.iter().any(|p| p == Protocol::P2pCircuit); // TODO: Make this an extension on `Multiaddr`.
```
Here too.

thomaseizinger committed:
Reduce diff

thomaseizinger commented on swarm/src/lib.rs:
```diff
@@ -1608,6 +1716,9 @@ pub enum DialError {
         obtained: PeerId,
         endpoint: ConnectedPoint,
     },
+    Denied {
+        cause: ConnectionDenied,
+    },
```
The addition of this variant makes this a breaking change.

thomaseizinger commented on swarm/src/lib.rs:
```diff
@@ -1705,8 +1820,13 @@ pub enum ListenError {
+    LocalPeerId {
+        endpoint: ConnectedPoint,
+    },
+    Denied {
+        cause: ConnectionDenied,
+    },
```
Here too.

thomaseizinger committed:
Fix rustdoc

thomaseizinger committed:
One more missing rustdoc link

thomaseizinger committed:
Merge branch '2824-no-handler-in-dial' into 2824-deprecate-into-connection-handler

thomaseizinger committed:
Remove `banned_peer_connections`

We no longer add connections for banned peers to the pool so we don't
have to filter them later.

thomaseizinger committed:
Fix test

Now that we no longer add connections to banned peers, we need to:

- Reduce the number of expected connections by 1 (i.e. delete line 1954)
- Act in `Stage::BannedDial` earlier (i.e. remove the `if`)

thomaseizinger committed:
Remove duplication around detecting relayed addresses in `libp2p-relay`

thomaseizinger committed:
Extract utility fn for dcutr protocol

thomaseizinger committed:
Remove unused import

thomaseizinger committed:
Remove underscore from used variable

thomaseizinger committed:
Merge branch 'master' into 2824-no-handler-in-dial

thomaseizinger committed:
Add changelog entry

thomaseizinger committed:
Add footnote

thomaseizinger committed:
Don't use footnotes

thomaseizinger committed:
Merge branch 'master' into 2824-no-handler-in-dial

thomaseizinger committed:
Inline `spawn` function

thomaseizinger committed:
Box future inside `spawn`

thomaseizinger committed:
Merge impl blocks

thomaseizinger committed:
Gracefully close connection to banned peer

thomaseizinger committed:
Merge branch '2824-no-handler-in-dial' into 2824-deprecate-into-connection-handler

thomaseizinger committed:
Ensure new connections are either spawned or gracefully closed

thomaseizinger committed:
Ignore error when closing connection

thomaseizinger committed:
Merge branch '2824-no-handler-in-dial' into 2824-deprecate-into-connection-handler

thomaseizinger commented on swarm/src/connection/pool.rs:
```diff
@@ -812,6 +819,48 @@ where
+///
+/// This connection has just been established but isn't part of the [`Pool`] yet.
+/// It either needs to be spawned via [`Pool::spawn_connection`] or dropped if undesired.
+///
+/// On drop, this type send the connection back to the [`Pool`] where it will be gracefully closed.
+#[derive(Debug)]
+pub struct NewConnection {
+    connection: Option<StreamMuxerBox>,
+    drop_sender: Option<oneshot::Sender<StreamMuxerBox>>,
+}
+
+impl NewConnection {
+    fn new(conn: StreamMuxerBox) -> (Self, oneshot::Receiver<StreamMuxerBox>) {
+        let (sender, receiver) = oneshot::channel();
+
+        (
+            Self {
+                connection: Some(conn),
+                drop_sender: Some(sender),
+            },
+            receiver,
+        )
+    }
+
+    fn extract(mut self) -> StreamMuxerBox {
+        self.connection.take().unwrap()
+    }
+}
+
+impl Drop for NewConnection {
+    fn drop(&mut self) {
+        if let Some(connection) = self.connection.take() {
+            let _ = self
+                .drop_sender
+                .take()
+                .expect("`drop_sender` to always be `Some`")
+                .send(connection);
+        }
+    }
+}
```
@mxinden: I think this should solve your "always gracefully close connections" concern :)

mxinden replied:
At first I thought this is too implicit for an internal component (i.e. `Pool`). Thinking about it some more, it does simplify error handling in `Swarm`. I am in favor of it. Thanks!

thomaseizinger committed:
Merge branch 'master' into 2824-deprecate-into-connection-handler

thomaseizinger committed:
Fix compile error

mxinden commented on swarm/CHANGELOG.md:
```diff
@@ -33,6 +33,27 @@
+  - `NetworkBehaviour::handle_established_outbound_connection`
+
+  Please note that due to limitations in the Rust compiler, _implementations_ of `new_handler` and `addresses_of_peer` are not flagged as deprecated.
+  Nevertheless, they will be removed in the future.
```
What do you think of linking to this from the main `CHANGELOG.md` file as a special callout for the release?

mxinden commented on swarm/src/handler/either.rs:
```diff
@@ -19,84 +19,23 @@
-pub enum IntoEitherHandler<L, R> {
-    Left(L),
-    Right(R),
-}
```
Given that `IntoConnectionHandler` is only deprecated at this point, should this already be removed or only removed once we remove `IntoConnectionHandler`?

thomaseizinger replied:
Yeah we can retain it!

mxinden commented on swarm/src/lib.rs:
```diff
@@ -756,15 +784,76 @@ where
                 established_in,
             } => {
                 if self.banned_peers.contains(&peer_id) {
-                    self.pool.close_connection(connection);
```
Following up on https://github.com/libp2p/rust-libp2p/pull/3328#discussion_r1101991956 and https://github.com/libp2p/rust-libp2p/pull/3254/#pullrequestreview-1292581602 it was not intuitive for me on a first read that `connection` on drop would close the underlying connection.

Thinking about it some more, I guess it is fine and significantly simplifies the error handling in this match arm.

thomaseizinger replied:
Yes, I figured doing it on `Drop` is easier given that we would have several paths in here that require the connection to be closed properly. I see it as the responsibility of `Pool` to do this, meaning it should provide a hard-to-misuse API, even if it is just an internal one.

mxinden commented on swarm/src/lib.rs:
```diff
@@ -802,6 +891,7 @@ where
                             other_established: other_established_connection_ids.len(),
                         },
                     ));
+                self.supported_protocols = supported_protocols;
```
This is new behavior, right? Previously we would only retrieve the supported protocols once at start-up. Now we retrieve them on each new connection. Could this cause any side-effects for folks depending on the previous behavior? Would you mind introducing this change in a separate pull request?

thomaseizinger replied:
I don't think we can do this in a separate PR. It definitely wasn't possible in earlier iterations. Will check.

The problem is that `new_handler` will panic by default now because you are not meant to call it. Thus, I cannot rely on it anywhere but have to call the new callbacks.

Those require connection-specific metadata that I only have after one is established.

thomaseizinger replied:
> Could this cause any side-effects for folks depending on the previous behavior? Would you mind introducing this change in a separate pull request?

Previously, we gathered the supported protocols on start-up of the swarm. Now, we are gathering it on the first connection. I think this can be observable if a `NetworkBehaviour` expects to read all supported protocols on the first call to `NetworkBehaviour::poll` via `PollParameters` when no connection is established yet.

I don't see a way around this change though unless we want to first completely redesign how `NetworkBehaviour`s signal which protocols they support.

We already agreed that `PollParameters` needs to go away (https://github.com/libp2p/rust-libp2p/issues/3124). I think the answer to this is that as part of https://github.com/libp2p/rust-libp2p/issues/2680, we scope supported protocols to a connection and thus remove this Swarm-wide state of "supported protocols".

mxinden replied:
Thanks for elaborating. Let's keep as is :+1:

mxinden commented on swarm/src/lib.rs:
```diff
@@ -1733,11 +1853,37 @@ impl error::Error for ListenError {
 
+#[derive(Debug)]
+pub struct ConnectionDenied {
+    inner: Box<dyn error::Error + Send + Sync + 'static>,
```
I am assuming that we require `Send` here to allow users to forward this error through a channel at a later point?

Why does the error need to be `Sync`?

thomaseizinger replied:
I think it is because we pass references of the error to the behaviours via `FromSwarm`.

thomaseizinger replied:
If the error is not `Sync`, users won't be able to use this with `anyhow`. See:

```
error[E0277]: `(dyn std::error::Error + std::marker::Send + 'static)` cannot be shared between threads safely
   --> interop-tests/src/bin/ping.rs:115:48
    |
115 |         swarm.dial(other.parse::<Multiaddr>()?)?;
    |                                                ^ `(dyn std::error::Error + std::marker::Send + 'static)` cannot be shared between threads safely
    |
```

mxinden commented on swarm/src/lib.rs:
```diff
@@ -1891,10 +2037,8 @@ mod tests {
-
         // Since the test does not try to open any substreams, we can
-        // use keep alive protocols handler.
+        // use the dummy protocols handler.
```
Merge conflict with https://github.com/libp2p/rust-libp2p/pull/3465.

mxinden commented on swarm/src/lib.rs:
```diff
@@ -1953,31 +2096,25 @@ mod tests {
                     }
                 }
                 Stage::BannedDial => {
-                    if swarm1.behaviour.assert_disconnected(s1_expected_conns, 2) {
```
Merge conflict with https://github.com/libp2p/rust-libp2p/pull/3465.

mxinden commented on protocols/relay/src/priv_client/transport.rs:
```diff
@@ -247,7 +248,7 @@ struct RelayedMultiaddr {
 
 /// Parse a [`Multiaddr`] containing a [`Protocol::P2pCircuit`].
 fn parse_relayed_multiaddr(addr: Multiaddr) -> Result<RelayedMultiaddr, TransportError<Error>> {
-    if !addr.iter().any(|p| matches!(p, Protocol::P2pCircuit)) {
+    if !addr.is_relayed() {
```
:rocket:

mxinden commented on protocols/kad/src/behaviour.rs:
```diff
@@ -1978,21 +1980,66 @@ impl<TStore> NetworkBehaviour for Kademlia<TStore>
-    fn new_handler(&mut self) -> Self::ConnectionHandler {
-        KademliaHandlerProto::new(KademliaHandlerConfig {
-            protocol_config: self.protocol_config.clone(),
-            allow_listening: true,
```
Uuf, when did we miss this?

thomaseizinger replied:
What do you mean?

mxinden replied:
If I understand correctly `allow_listening` is always `true`, thus a dummy configuration option, thus fine to remove. I am assuming that this is a relict from a past refactoring. Not directly related to this pull request. Resolving the comment.

thomaseizinger replied:
It is not removed in this PR but I guess we should remove it, I'll open an issue!

mxinden commented on protocols/dcutr/src/behaviour_impl.rs:
```diff
@@ -386,3 +460,7 @@ impl NetworkBehaviour for Behaviour {
     }
 }
+
+fn is_relayed(addr: &Multiaddr) -> bool {
+    addr.iter().any(|p| p == Protocol::P2pCircuit)
+}
```
No extension trait? Not a strong opinion.

thomaseizinger replied:
Didn't think it was worth it if the usage is only in a single file.

mxinden commented on protocols/dcutr/src/behaviour_impl.rs:
```diff
@@ -237,11 +239,82 @@ impl Behaviour {
-        handler::Prototype
+    fn handle_established_inbound_connection(
+        &mut self,
+        _connection_id: ConnectionId,
```
Why the `_`? It is used further below in the `remove` call.

thomaseizinger replied:
Copy-paste error 🙃

thomaseizinger replied:
I opened https://github.com/libp2p/rust-libp2p/pull/3484 to prevent this in the future.

mxinden commented on protocols/dcutr/src/behaviour_impl.rs:
```diff
@@ -237,11 +239,82 @@ impl Behaviour {
+                    Either::Left(handler::relayed::Handler::new(ConnectedPoint::Listener {
+                        local_addr: local_addr.clone(),
+                        send_back_addr: remote_addr.clone(),
+                    })) // TODO: We could make two `handler::relayed::Handler` here, one inbound one outbound.
```
Can you expand on this comment?

thomaseizinger replied:
"dcutr" cares about who the dialer and listener on a relayed connection is. Now that we have different callbacks for this, it is even more obvious that we could split this logic into two handlers instead of being conditional on the `ConnectedPoint` within the single handler. That would remove some `panic!`s in the implementation that are unreachable.

mxinden reviewed (commented):
Overall looking great. I have a couple of comments and questions. Nothing major.

thomaseizinger committed:
Restore `IntoEitherHandler`

thomaseizinger committed:
Don't underscore used variable

thomaseizinger committed:
Merge branch 'master' into 2824-deprecate-into-connection-handler

thomaseizinger committed:
Undo merge conflicts with #3465

thomaseizinger committed:
Call out limitations in Rust compiler

thomaseizinger committed:
Move manage conn entry to top of changelog

This one is more important.

thomaseizinger committed:
Fix fmt

mxinden reviewed (approved):
Ready for merge from my end. Please add the `send-it` label once you want this merged @thomaseizinger. Including it in v0.51.0 sounds good to me.

thomaseizinger committed:
Merge branch 'master' into 2824-deprecate-into-connection-handler

# Conflicts:
#	swarm/CHANGELOG.md
